src/it/scala/net/snowflake/spark/snowflake/SnowflakeTelemetrySuite.scala

```diff
@@ -162,12 +173,197 @@ class SnowflakeTelemetrySuite extends IntegrationSuiteBase {
     }
   }
 
-  private def nodeContains(node: JsonNode, value: String): Boolean = {
-    val iterator = node.iterator()
-    while (iterator.hasNext) {
-      if (iterator.next().asText().equals(value)) return true
+  test("IT test: PLAN_STATISTIC: read from snowflake") {
+    // Enable dummy sending telemetry message.
+    val messageBuffer = mutable.ArrayBuffer[ObjectNode]()
+    val oldSender = SnowflakeTelemetry.setTelemetryMessageSenderForTest(
+      new MockTelemetryMessageSender(messageBuffer))
+    try {
+      // A basis dataframe read
+      val df1 = sparkSession.read
+        .format(SNOWFLAKE_SOURCE_NAME)
+        .options(connectorOptionsNoTable)
+        .option("query", "select 123 as A")
+        .load()
+      df1.collect()
+
+      val planStatisticMessages = messageBuffer
+        .filter(_.get("type").asText().equals("spark_plan_statistic"))
+      assert(planStatisticMessages.nonEmpty)
+      planStatisticMessages.foreach { x =>
+        val planStatistics = x.get("data").get(TelemetryFieldNames.STATISTIC_INFO)
+        assert(planStatistics.isArray && planStatistics.size() > 0)
+        assert(nodeContains(planStatistics, "LogicalRelation:SnowflakeRelation"))
+        assert(!nodeContains(planStatistics, "SaveIntoDataSourceCommand:DefaultSource"))
+      }
+    } finally {
+      // Reset to the real Telemetry message sender
+      SnowflakeTelemetry.setTelemetryMessageSenderForTest(oldSender)
+    }
+  }
+
+  test("IT test: PLAN_STATISTIC: read from snowflake and write to snowflake") {
+    // Enable dummy sending telemetry message.
+    val messageBuffer = mutable.ArrayBuffer[ObjectNode]()
+    val oldSender = SnowflakeTelemetry.setTelemetryMessageSenderForTest(
+      new MockTelemetryMessageSender(messageBuffer))
+    try {
+      // A basis dataframe read
+      val df1 = sparkSession.read
+        .format(SNOWFLAKE_SOURCE_NAME)
+        .options(connectorOptionsNoTable)
+        .option("query", "select current_timestamp()")
+        .load()
+
+      df1.write
+        .format(SNOWFLAKE_SOURCE_NAME)
+        .options(connectorOptionsNoTable)
+        .option("dbtable", test_table)
+        .mode(SaveMode.Overwrite)
+        .save()
+
+      val planStatisticMessages = messageBuffer
+        .filter(_.get("type").asText().equals("spark_plan_statistic"))
+        .filter { x =>
+          val planStatistics = x.get("data").get(TelemetryFieldNames.STATISTIC_INFO)
+          assert(planStatistics.isArray && planStatistics.size() > 0)
+          nodeContains(planStatistics, "LogicalRelation:SnowflakeRelation") &&
+            nodeContains(planStatistics, "SaveIntoDataSourceCommand:DefaultSource")
+        }
+      assert(planStatisticMessages.length == 1)
```

Comment:
It seems there is a lot of duplicated code for every test case. Can we move the common logic into a method?

Reply:
it can be extracted to a separate function. But I prefer not to do that. This is a test case, I think the easy-test-case-failure analysis is more important than code duplication.